src/hotspot/share/runtime/cpuTimeCounters.cpp

```diff
@@ -0,0 +1,130 @@
+#include "runtime/atomic.hpp"
+#include "runtime/os.hpp"
+
+const char* CPUTimeGroups::to_string(CPUTimeType val) {
+  switch (val) {
+    case CPUTimeType::gc_total:
+      return "total_gc_cpu_time";
+    case CPUTimeType::gc_parallel_workers:
+      return "gc_parallel_workers";
+    case CPUTimeType::gc_conc_mark:
+      return "gc_conc_mark";
+    case CPUTimeType::gc_conc_refine:
+      return "gc_conc_refine";
+    case CPUTimeType::gc_service:
+      return "gc_service";
+    case CPUTimeType::vm:
+      return "vm";
+    case CPUTimeType::conc_dedup:
+      return "conc_dedup";
+    default:
+      ShouldNotReachHere();
+  };
+}
+
+bool CPUTimeGroups::is_gc_counter(CPUTimeType val) {
+  switch (val) {
+    case CPUTimeType::gc_parallel_workers:
+      return true;
+    case CPUTimeType::gc_conc_mark:
+      return true;
+    case CPUTimeType::gc_conc_refine:
+      return true;
+    case CPUTimeType::gc_service:
+      return true;
+    default:
+      return false;
+  }
+  ShouldNotReachHere();
+}
+
+CPUTimeCounters* CPUTimeCounters::_instance = nullptr;
+
+CPUTimeCounters::CPUTimeCounters() :
+    _cpu_time_counters(),
+    _gc_total_cpu_time_diff(0) {
+  create_counter(SUN_THREADS, CPUTimeGroups::CPUTimeType::gc_total);
+}
+
+void CPUTimeCounters::inc_gc_total_cpu_time(jlong diff) {
+  Atomic::add(&_gc_total_cpu_time_diff, diff);
+}
+
+void CPUTimeCounters::publish_gc_total_cpu_time() {
+  // Ensure that we are only incrementing atomically by using Atomic::cmpxchg
+  // to set the value to zero after we obtain the new CPU time difference.
+  jlong old_value;
+  jlong fetched_value = Atomic::load(&_gc_total_cpu_time_diff);
+  jlong new_value = 0;
+  do {
+    old_value = fetched_value;
+    fetched_value = Atomic::cmpxchg(&_gc_total_cpu_time_diff, old_value, new_value);
+  } while (old_value != fetched_value);
+  get_counter(CPUTimeGroups::CPUTimeType::gc_total)->inc(fetched_value);
+}
+
+void CPUTimeCounters::create_counter(CounterNS ns, CPUTimeGroups::CPUTimeType name) {
+  if (UsePerfData) {
+    EXCEPTION_MARK;
+    if (os::is_thread_cpu_time_supported()) {
+      _cpu_time_counters[static_cast<int>(name)] =
+                  PerfDataManager::create_counter(ns, CPUTimeGroups::to_string(name),
+                                                  PerfData::U_Ticks, CHECK);
+    }
+  }
+}
+
+void CPUTimeCounters::create_counter(CPUTimeGroups::CPUTimeType group) {
+  create_counter(SUN_THREADS_CPUTIME, group);
+}
+
+PerfCounter* CPUTimeCounters::get_counter(CPUTimeGroups::CPUTimeType name) {
+  return _cpu_time_counters[static_cast<int>(name)];
+}
+
+ThreadTotalCPUTimeClosure::~ThreadTotalCPUTimeClosure() {
+    CPUTimeCounters* instance = CPUTimeCounters::get_instance();
+    PerfCounter* counter = instance->get_counter(_name);
+    jlong net_cpu_time = _total - counter->get_value();
+    counter->inc(net_cpu_time);
+    if (CPUTimeGroups::is_gc_counter(_name)) {
+      instance->inc_gc_total_cpu_time(net_cpu_time);
+    }
```

Comment:
I feel much of this is on the wrong abstraction level; `CPUTimeCounters::update_counter(_name, _total);` should be sufficient here. (The caller handles diff calculation and inc gc-counter if needed.)

Reply:
We could add a new closure just used by GC that 's a sub-class of `ThreadTotalCPUTimeClosure` and just adds this to the constructor: 
```
instance->inc_gc_total_cpu_time(net_cpu_time);
```

That way we could get rid of `CPUTimeGroups::is_gc_counter()` as well since all those counters should use the "GC closure" or we can keep it and assert that no GC closure uses the wrong closure.

What do you think about that Albert, would that address your concerns?

Reply:
(I just realized that I made a typo in my previous msg; should be *callee* instead.) That is what I have in mind.

```cpp
void CPUTimeCounters::update_counter(name, total) {
  auto counter = get_counter(name);
  auto old_v = counter->get_value();
  auto diff = total - old_v;
  counter->inc(diff);
  if (counter->is_gc_counter()) {
    counter->inc(diff);
  }
}
```

Reply:
I'm not sure I understood correctly, could you let me know if this latest commit addresses your comment in the way you were intending?

Reply:
It does.